extendr-api/src/wrapper/matrix.rs

```diff
@@ -159,53 +186,35 @@ trait Offset<D> {
     fn offset(&self, idx: D) -> usize;
 }
 
-impl<T> Offset<[usize; 1]> for RArray<T, [usize; 1]> {
-    /// Get the offset into the array for a given index.
-    fn offset(&self, index: [usize; 1]) -> usize {
-        if index[0] - BASE > self.dim[0] {
-            panic!("array index: row overflow");
-        }
-        index[0] - BASE
-    }
-}
-
-impl<T> Offset<[usize; 2]> for RArray<T, [usize; 2]> {
+impl<T, const NDIM: usize> Offset<[usize; NDIM]> for RArray<T, NDIM> {
     /// Get the offset into the array for a given index.
-    fn offset(&self, index: [usize; 2]) -> usize {
-        if index[0] - BASE > self.dim[0] {
-            panic!("matrix index: row overflow");
-        }
-        if index[1] - BASE > self.dim[1] {
-            panic!("matrix index: column overflow");
+    fn offset(&self, index: [usize; NDIM]) -> usize {
+        let dims = self.get_dim();
+        if index.len() != dims.len() {
+            panic!("array index: dimension mismatch");
         }
-        (index[0] - BASE) + self.dim[0] * (index[1] - BASE)
-    }
-}
-
-impl<T> Offset<[usize; 3]> for RArray<T, [usize; 3]> {
-    /// Get the offset into the array for a given index.
-    fn offset(&self, index: [usize; 3]) -> usize {
-        if index[0] - BASE > self.dim[0] {
-            panic!("RMatrix3D index: row overflow");
-        }
-        if index[1] - BASE > self.dim[1] {
-            panic!("RMatrix3D index: column overflow");
-        }
-        if index[2] - BASE > self.dim[2] {
-            panic!("RMatrix3D index: submatrix overflow");
-        }
-        (index[0] - BASE) + self.dim[0] * (index[1] - BASE + self.dim[1] * (index[2] - BASE))
+        index
+            .iter()
+            .zip(dims.iter())
+            .rev()
+            .enumerate()
+            .fold(0, |acc, (i, (&idx, &dim))| {
+                println!("idx: {}, dim: {}", idx, dim);
+                if idx - BASE >= dim {
+                    panic!("array index: dimension {} overflow (0-based dimension)", i);
```

Comment:
same as above. we should err for no panics and instead use `throw_r_error()` here

Reply:
Thanks, I'll update this